CSI Node info registration in kubelet

verult opened the pull request:
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #67683

**Special notes for your reviewer**:
Feature issue: https://github.com/kubernetes/features/issues/557
Design doc: https://github.com/kubernetes/community/pull/2034

Missing pieces:
* CSI client retry and exponential backoff logic.
* CSINodeInfo object validation
* e2e test with all the CSI machinery.

An RBAC rule is also added to support external-provisioner topology updates.

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Registers volume topology information reported by a node-level Container Storage Interface (CSI) driver. This enables Kubernetes support of CSI topology mechanisms.
```

verult commented:
/sig storage
/assign @liggitt @saad-ali @vladimirvivien @sbezverk @msau42

verult commented on pkg/volume/csi/nodeinfomanager/nodeinfomanager.go:
```diff
@@ -0,0 +1,384 @@
+
+// updateNode repeatedly attempts to update the corresponding node object
+// which is modified by applying the given update functions in order.
+func (lm *nodeInfoManager) updateNode(updateFuncs []nodeUpdateFunc) error {
```
Will rename all the receivers to `nim`.

verult commented on pkg/volume/csi/nodeinfomanager/nodeinfomanager.go:
```diff
@@ -0,0 +1,384 @@
+		}
+
+		if val, ok := existingDriverMap[csiDriverName]; ok {
+			if reflect.DeepEqual(val, topologyKeys) {
```
Will check set equality instead in the new version that uses `CSINodeInfo`.

verult commented:
Feel free to hold off on reviewing this PR for now - I'm almost done with wiring up CSINodeInfo and a good amount of logic will become obsolete

verult commented:
Added integration with `CSINodeInfo`. Please only review the last commit - everything before it are rebased from https://github.com/kubernetes/kubernetes/pull/67803.

Missing functionality is updated in the top post, but adding those won't cause any major structural changes. The PR is ready for review now.

@liggitt @saad-ali @vladimirvivien @sbezverk @msau42 @gnufied

msau42 commented on pkg/volume/csi/nodeinfomanager/nodeinfomanager.go:
```diff
@@ -0,0 +1,447 @@
+		// exponential backoff to avoid exhausting the apiserver.
+
+		nodeClient := nim.k8s.CoreV1().Nodes()
+		node, err := nodeClient.Get(string(nim.nodeName), metav1.GetOptions{})
```
are there plans to use a shared informer?  If so, we should clone the node object first before modifying it so it doesn't impact other uses that may be sharing the same pointer.

verult replied:
No I wasn't planning to. Driver registration is fairly infrequent so watching the object doesn't seem worth it

msau42 commented on pkg/volume/csi/nodeinfomanager/nodeinfomanager.go:
```diff
@@ -0,0 +1,447 @@
+
+		needUpdate := false
+		for _, update := range updateFuncs {
+			newNode, updated, err := update(node)
```
This works only if the multiple updateFuncs are not conflicting with each other.

verult replied:
I'll make it more explicit in the comments for this function that because updateFuncs are called in order, each updateFunc should consider the effects of previous updateFuncs

liggitt replied:
> each updateFunc should consider the effects of previous updateFuncs

what does that mean?

verult replied:
I added a comment here - does that help or should it be more clear?
https://github.com/kubernetes/kubernetes/blob/9c855fb9e62de46f511d9dc0446425c1b5a76499/pkg/volume/csi/nodeinfomanager/nodeinfomanager.go#L119

verult replied:
Well said! Will update the comment

msau42 commented on pkg/volume/csi/nodeinfomanager/nodeinfomanager.go:
```diff
@@ -0,0 +1,447 @@
+			driverInfo.NodeID = driverNodeID
+			driverInfo.TopologyKeys = topologyKeys
+		}
+		driverInfos = append(driverInfos, driverInfo)
```
Should we be adding all drivers in this call to just update a single driver?  Do all the CSI drivers register again when we upgrade the node?

Should we be looking at the annotations at all?  IIUC, this is supposed to be the replacement for the annotations

verult replied:
Yeah CSI drivers should register again upon node upgrade, and this is here as a safeguard to ensure annotations and CSINodeInfo are consistent while we support both. There is always a chance that kubelet is upgraded while the previous Node object stays, and annotations and CSINodeInfo could be momentarily out of sync.

msau42 replied:
Is it necessary to keep the annotations and CSIDriverInfo in sync on the first driver's registration, if the other driver will register again later?  In an upgrade scenario, then this means we're on the first boot after upgrade and we're processing the first driver registration.  We're adding an entry for all the other drivers too even though they have not gone through registration yet.  Could that cause other issues?

jsafrane replied:
I think this method should get whole list of registered CSI drivers so it creates/updates complete CSINodeInfo with all drivers. The result should be the same in the end, however it feels more robust. I don't line CSINodeInfo with empty topology for some drivers just because we translated it from an annotation.

msau42 replied:
This method gets called for a single driver registration though, and only on the first one since it's creating the object for the first time.

verult replied:
Discussed with @msau42 offline. The current implementation doesn't guarantee the node info is always up-to-date. In particular, with the current kubelet plugin registration design, there is no way to remove a non-existent driver unless it detects that a driver socket is removed (after unregistration is implemented). If a driver is removed while kubelet is down, the node information will remain. So whatever issue comes up when CSINodeInfo is kept in sync with annotation is unavoidable today.

But I'll remove the logic to keep the two sources in sync, because it complicates the implementation without adding too much benefit. One minor advantage of keeping them in sync is to avoid debugging confusion, but I think it'll be OK.

verult commented:
@saad-ali and I discussed offline, and we are going to leave the CSI label prefix as is in Kubernetes instead of reversing it, for the time being.

msau42 commented on pkg/volume/csi/csi_plugin.go:
```diff
@@ -103,42 +103,41 @@ func RegistrationCallback(pluginName string, endpoint string, versions []string,
+		// all other CSI components will be able to get the actual socket of CSI drivers by its name.
+		csiDrivers.Lock()
+		defer csiDrivers.Unlock()
+		csiDrivers.driversMap[pluginName] = csiDriver{driverName: pluginName, driverEndpoint: endpoint}
```
Previously the lock was being held across the entire function instead of just the map update.  Could there be any issues if this map has an entry but we haven't finished registering the plugin yet?  Should we update the map at the very end?

verult replied:
In both the previous and the current version, the lock is only held on write, all within this `RegistrationCallback()` function. The map has to be updated before the `csi.NodeGetInfo()` call unfortunately because it relies on the driver info to call out to gRPC.

However, when plugin unregistration is implemented in the future, depending on how it's implemented, there could be a race between calls of `RegistrationCallback()`. Will update to lock the entire function once.

verult replied:
Discussed with @msau42 offline. For now we'll keep the map locked as short as possible. When unregistration is implemented we'll reconsider again.

Will add comments explaining why it's OK to keep locking short.

verult committed:
Implemented logic in kubelet for registering node info, including wiring to CSINodeInfo; added unit tests for node updates; updated RBAC, NodeAuthorizer, NodeRestriction.

verult commented on pkg/volume/csi/nodeinfomanager/nodeinfomanager.go:
```diff
@@ -0,0 +1,447 @@
+		return fmt.Errorf("error adding CSI driver node info: driverNodeID must not be empty")
+	}
+
+	if topology == nil {
```
Yeah, thanks for the catch!

verult commented:
Resolved comments

jsafrane commented on pkg/volume/csi/nodeinfomanager/nodeinfomanager.go:
```diff
@@ -0,0 +1,450 @@
+
+// AddNodeInfo updates the node ID annotation in the Node object and CSIDrivers field of the
+// CSINodeInfo object.
+func (nim *nodeInfoManager) AddNodeInfo(driverName string, driverNodeID string, topology *csipb.Topology) error {
```
Should it be locked somehow? Is it possible that two drivers gets registered at the same time and AddNodeInfo is called twice?

verult replied:
Yep it's entirely possible 2 drivers are registered at the same time - socket fsnotify events are handled in parallel.

None of the node info manger's methods mutate any internal state. However multiple calls will cause retries due to Node/CSINodeInfo conflict. Even though kubelet ends up making more calls than needed, it might be safer than using locks, which could potentially block progress due to network calls.

I'll call this out in a comment

verult replied:
However in the future when plugin unregistration is implemented, there could be a race between `AddNodeInfo()` and `RemoveNodeInfo()` on creating and deleting `CSINodeInfo`...

verult replied:
We could cross that bridge when we get there, since plugin unregistration implementation might already have mechanisms in place to serialize register and unregister calls.

saad-ali commented on pkg/volume/csi/nodeinfomanager/nodeinfomanager.go:
```diff
@@ -0,0 +1,461 @@
+
+	err = nim.updateCSINodeInfo(driverName, driverNodeID, topology)
+	if err != nil {
+		return fmt.Errorf("error updating CSINodeInfo object with CSI driver node info: %v", err)
```
Maybe instead of failing, just log a warning for now saying you could not update the object and that topology may not work (caller has to enable CRD in order for it to work). And leave a TODO, once CSINodeInfo object goes beta you can fail instead of log?

verult replied:
ACK will update. Currently if an error is returned, the registration handler proceeds to unregister the driver, and we definitely don't wan this if CRD creation is disabled

verult replied:
It would have to move to GA right, since a user can still disable the CRD in beta?

verult commented:
Addressed comments; rebased to the latest version of Cluster Registry API PR (https://github.com/kubernetes/kubernetes/pull/67803) to minimize conflicts.

Given limited time I'll drop the remaining tasks in the PR for now.
* Retry is not core to functionality
* Validation is going to require external validation webhooks since OpenAPI validation on k8s doesn't support uniqueItems (for set validation) nor map key validation.
* I think we need better e2e testing for CSI testing in k8s in general, and it's coming

verult commented:
Placing hold until #67803 merges

/hold

liggitt commented on plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go:
```diff
@@ -160,11 +160,16 @@ func NodeRules() []rbacv1.PolicyRule {
 	}
+	if utilfeature.DefaultFeatureGate.Enabled(features.KubeletPluginsWatcher) {
+		csiNodeInfoRule := rbacv1helpers.NewRule("get", "watch", "list", "update", "patch", "create").Groups("csi.storage.k8s.io").Resources("csinodeinfos").RuleOrDie()
+		nodePolicyRules = append(nodePolicyRules, csiNodeInfoRule)
```
This lets any node update the CSI info for any other node. I expected nodes to only be able to update their own CSI info. That is typically accomplished via changes in the node authorizer and noderestriction admission plugins. 

@mtaufen for pointers from your recent node lease change

@mikedanese fyi

verult replied:
Should the Node Restriction rule be added *in addition to* the RBAC rule here, in case NodeRestriction admission plugin is disabled?

verult replied:
Looks like the authorizer uses internal APIs for objects it authorizes ([link](https://github.com/kubernetes/kubernetes/blob/ef741c3038bb26326daed17ac722cce8843b8f0d/plugin/pkg/auth/authorizer/node/node_authorizer.go#L30)). For `CSINodeInfo` this has to be external, i.e. "k8s.io/csi-api/pkg/apis/csi/v1alpha1". Would this cause any issues?

liggitt replied:
No, only resource and API group are required to authorize. Version is not used.

liggitt replied:
> Are there any special considerations if the object is a CRD?

If the authorization and admission restriction is purely based on the name of the incoming object, then no. If watching the existing objects with an informer was required, that might be more involved, but iirc, the intent was to have the name of the csi info object match the node name, right?

verult replied:
> but iirc, the intent was to have the name of the csi info object match the node name, right?
Yep that's correct, and we don't use an informer here.

For the external provisioner, it's sufficient to have just the RBAC read rules, since it needs to read from all CSINodeInfo objects, right?

verult commented:
Added RBAC, NodeAuthorizer, and NodeRestriction updates

verult commented:
/unhold

verult commented:
/hold cancel

msau42 commented on pkg/volume/csi/nodeinfomanager/nodeinfomanager.go:
```diff
@@ -0,0 +1,450 @@
+	if err != nil {
+		// Can't return error because the driver will be unregistered, and the CSINodeInfo CRD may not
+		// be installed
+		// TODO return an error instead once CRD installation moves to GA.
```
Is the issue that if the the feature gate is not turned on, then this will fail?  If so, shouldn't we check for the feature gate here?

verult commented:
There is a follow-up PR to the cluster registry API PR: https://github.com/kubernetes/kubernetes/pull/68159

/hold

verult commented:
Tried to create the CRD at test initialization but got the error `the server could not find the requested resource`. Is the CRD object not installed in the integration test apiserver by default?

verult commented:
Yep looks like the APIExtensionsServer is not installed:
https://github.com/kubernetes/kubernetes/blob/e0782b99f1ac34bac5ec393abdf7455c16dd30ef/test/integration/framework/master_utils.go#L182

I could imagine doing something like
```go
masterConfig.Complete().New(apiextensionsConfig.Complete().New(genericapiserver.NewEmptyDelegate()))
```
but I'm not sure what's the right way, since the CRD test does something entirely different:
https://github.com/kubernetes/kubernetes/blob/e0782b99f1ac34bac5ec393abdf7455c16dd30ef/test/integration/master/crd_test.go#L112

@liggitt worst case could we leave out the integration test for the time being?

liggitt commented:
> @liggitt worst case could we leave out the integration test for the time being?

that doesn't seem proper. a feature we can't test well is not one we should merge. the integration test ensures the authorizer and admission bits work together as expected. unit tests only cover one component or the other.

as an aside, I think this reveals some of the problems in making in-tree components build on top of API types that are not built-in. I expect there will be more issues like this as this feature progresses.

verult commented:
ACK

> as an aside, I think this reveals some of the problems in making in-tree components build on top of API types that are not built-in. I expect there will be more issues like this as this feature progresses.
Agreed

saad-ali commented:
/milestone v1.12

saad-ali reviewed (approved):
mod node authorizer and integration tests changes which @liggitt is awaiting for
/lgtm
/approve

verult commented:
NodeAuthorizer and NodeRestriction updated, @liggitt PTAL

verult commented:
Rebased off of  #68159. Will remove hold once that merges

verult commented:
/retest

saad-ali commented:
/lgtm
/approve

liggitt commented on plugin/pkg/admission/noderestriction/admission.go:
```diff
@@ -143,6 +145,12 @@ func (c *nodePlugin) Admit(a admission.Attributes) error {
 		return admission.NewForbidden(a, fmt.Errorf("disabled by feature gate %s", features.NodeLease))
 
+	case csiNodeInfoResource:
+		if c.features.Enabled(features.CSICRDAutoInstall) {
```
Is `CSICRDAutoInstall` the right feature to check for this? This shouldn't really care how the CRD is installed, just if the node is using it

liggitt commented on plugin/pkg/auth/authorizer/node/node_authorizer.go:
```diff
@@ -120,7 +122,13 @@ func (r *NodeAuthorizer) Authorize(attrs authorizer.Attributes) (authorizer.Deci
 			}
 			return authorizer.DecisionNoOpinion, fmt.Sprintf("disabled by feature gate %s", features.NodeLease), nil
+		case csiNodeInfoResource:
+			if r.features.Enabled(features.CSICRDAutoInstall) {
```
Same question, this seems like the wrong feature to be checking

liggitt commented on plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go:
```diff
@@ -502,6 +496,22 @@ func ClusterRoles() []rbacv1.ClusterRole {
+		rbacv1helpers.NewRule("get", "list", "watch").Groups(legacyGroup).Resources("nodes").RuleOrDie(),
+	}
+	if utilfeature.DefaultFeatureGate.Enabled(features.KubeletPluginsWatcher) {
+		externalProvisionerRules = append(externalProvisionerRules, rbacv1helpers.NewRule("get", "watch", "list").Groups("csi.storage.k8s.io").Resources("csinodeinfos").RuleOrDie()) // TODO (verult) remove get permission?
```
If list and watch are present, get is fine as well

liggitt commented:
I was surprised to see api changes that didn't get labeled as api changes or require api approval. Shouldn't 
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/csi-api/OWNERS follow normal api approval procedures (sig and subject matter experts are tagged for reviews, api approvers for approval?)

verult commented on plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go:
```diff
@@ -160,6 +160,10 @@ func NodeRules() []rbacv1.PolicyRule {
 		volAttachRule := rbacv1helpers.NewRule("get").Groups(storageGroup).Resources("volumeattachments").RuleOrDie()
 		nodePolicyRules = append(nodePolicyRules, volAttachRule)
 	}
+	if utilfeature.DefaultFeatureGate.Enabled(features.KubeletPluginsWatcher) {
```
Kubelet only talks to CSINodeInfo if `KubeletPluginsWatcher` is enabled.

The logic also handles the the scenario where CRD is not installed gracefully.

verult commented:
@liggitt are you referring to the changes introduced in the first 3 commits? Those are temporarily rebased from #68159

liggitt commented:
> @liggitt are you referring to the changes introduced in the first 3 commits? Those are temporarily rebased from #68159

Ok, glad to see API folks tagging there. Nevertheless, I'd expect them to have shown up as api changes in this pull as well. I think https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/csi-api/pkg/apis probably needs an OWNERS file that routes to csi reviewers for review and api-approvers for approval

verult commented:
ACK. @saad-ali FYI

verult commented:
@liggitt addressed your comments, PTAL

saad-ali commented:
/priority critical-urgent
/lgtm

liggitt commented on plugin/pkg/admission/noderestriction/admission.go:
```diff
@@ -146,10 +147,10 @@ func (c *nodePlugin) Admit(a admission.Attributes) error {
 
 	case csiNodeInfoResource:
-		if c.features.Enabled(features.CSICRDAutoInstall) {
+		if c.features.Enabled(features.KubeletPluginsWatcher) {
```
this seems more likely to be correct, but want eyes on this from the storage reviewers - is this the thing that controls whether kubelets make use of the CSINodeInfo API? using this here also means this gate will need to be set for the apiserver.

saad-ali replied:
This looks right to me too. Only when plugin registration is enabled do we attempt to write this object.

liggitt replied:
ok. be sure to document the feature gate is required on the apiserver

msau42 replied:
My understanding of the feature gates are:

- CSICRDAutoInstall: gates the CSINodeInfo object
- KubeletPluginsWatcher: gates the feature to have kubelet register plugins instead of the sidecar

Should we check for both feature gates in this PR?

verult replied:
Yup, the logic for kubelet-based driver registration is only triggered if the overall plugin registration system is enabled

saad-ali replied:
I'd like to avoid a proliferation of flags, so no go on a third flag. If Kubelet Plugin Registration is enabled. This is the way to do registration -- makes sense to me to couple it to that flag. I also thought about checking both flags but it seems to me this particular code shouldn't really care if the CRD is installed or not.

msau42 replied:
how do we reconcile KubeletPluginsWatcher being beta, and this being alpha?

verult replied:
The way I see it is `CSICRDAutoInstall` is a guarantee that k8s will try its best to make sure the CRD is installed. It's not necessarily a gate on the CSINodeInfo object itself.

@msau42 if by reconciling you mean how we handle the fact that `KubeletPluginsWatcher` is enabled by default and `CSICRDAutoInstall` is disabled by default, the update logic `AddNodeInfo()` should handle it gracefully.

liggitt replied:
> I think KubeletPluginsWatcher may be moving to beta

@msau42 if enabling that causes kubelets to start updating their labels, that's problematic. today, kubelets only label on initial registration. xref https://github.com/kubernetes/kubernetes/pull/68267 (which I thought could wait until early in 1.13 to go in, given the kublet is not updating its own labels currently):

> *Use cases like device plugins and CSI drivers intend to use self-added labels to express topology, which is fine, but those use cases need to be coordinated with controls the cluster-admin can use to ensure labels remain useful for partitioning a cluster along security-related boundaries (rather than starting to have kubelets set lots of labels and trying to wrap policy around that later).*

if we ship a release in which kubelets update their own labels unrestricted, our hands are tied for a year for closing the current gap to require cluster-admin opt-in for the labels they want to allow.

liggitt commented on plugin/pkg/admission/noderestriction/admission.go:
```diff
@@ -435,11 +436,11 @@ func (c *nodePlugin) admitCSINodeInfo(nodeName string, a admission.Attributes) e
-		nodeInfo, ok := a.GetObject().(*csiv1alpha1.CSINodeInfo)
-		if !ok {
-			return admission.NewForbidden(a, fmt.Errorf("unexpected type %T", a.GetObject()))
+		accessor, err := meta.Accessor(a.GetObject())
```
looks good

msau42 commented on pkg/volume/csi/nodeinfomanager/nodeinfomanager.go:
```diff
@@ -101,11 +101,11 @@ func (nim *nodeInfoManager) AddNodeInfo(driverName string, driverNodeID string,
 	if err != nil {
 		if utilfeature.DefaultFeatureGate.Enabled(features.CSICRDAutoInstall) {
```
This function both creates (if it doesn't exist) or updates the CSINodeInfo object.  I think the feature gate check should go around the `updateCSINodeInfo`.  We shouldn't even try if the feature is not enabled.

liggitt replied:
CSICRDAutoInstall should not mean the same thing as "use the CSINodeInfo API". Whether the CRD is auto-installed or not is orthogonal to whether the kubelet should use it.

verult replied:
Alternatively, we could *always* log instead of error, i.e. if the CRD doesn't exist, always fall back to relying on annotations only

msau42 replied:
I'm not a fan of falling back because then topology won't work when it's supposed to be working.

liggitt commented on pkg/volume/csi/nodeinfomanager/nodeinfomanager.go:
```diff
@@ -0,0 +1,455 @@
+		return fmt.Errorf("error adding CSI driver node info: driverNodeID must not be empty")
+	}
+
+	err := nim.updateNode(
```
is this intending to modify labels on the Node object using the kubelet's credentials?

verult replied:
Yeah

verult commented:
Squashed; haven't changed the flag gate check in `AddNodeInfo()` yet since it's still under discussion

liggitt commented:
the mechanics of the auth changes in this PR look good to me.

the feature flag used to enable the CSINodeInfo authz changes and the new self-labeling in nodeinfomanager should be one that will not be enabled by default in 1.12:
* the CSINodeInfo API is still alpha
* the self-labeling is not yet under policy control (xref https://github.com/kubernetes/kubernetes/pull/68267)

I agree with @saad-ali on avoiding a proliferation of flags. Would it make sense to rename the alpha `CSICRDAutoInstall` flag to `CSINodeInfo` and use it to limit:
* authz rules allowing kubelets to modify CSINodeInfo objects
* the new nodeinfomanager code in this PR (specifically node self-labeling)
* auto-install of the CSINodeInfo CRD

verult commented:
To be clear, prior to this PR kubelet already has the functionality to add CSI Node ID as a Node annotation. Since this logic was under the alpha `KubeletPluginsWatcher` gate, do we have the option to keep it under a different alpha gate for 1.12? If so, I propose we keep this annotation logic under whatever alpha gate we end up choosing as well to keep all CSI node update logic together.

liggitt commented:
>  Since this logic was under the alpha KubeletPluginsWatcher gate, do we have the option to keep it under a different alpha gate for 1.12? 

Until it is out of alpha, all behavior is free to move or be removed as makes sense. 

> If so, I propose we keep this annotation logic under whatever alpha gate we end up choosing as well to keep all CSI node update logic together.

As someone less familiar with the domain, that makes more sense to me.

verult commented:
CRD PR merged
/hold cancel

verult commented:
/kind feature

msau42 commented:
[CSICRDAutoInstall](https://github.com/kubernetes/kubernetes/pull/67803) currently only controls the auto-installation of the crd objects, but does not gate the creation of the CRDs themselves (is there a way to gate creation of alpha crds?). Is that a problem?

@verult are you suggesting that we don't move the Node annotation logic to beta, and keep it under some CSINodeRegistration alpha gate?

verult commented:
> Cheng Xing are you suggesting that we don't move the Node annotation logic to beta, and keep it under some CSINodeRegistration alpha gate?
Yup

msau42 commented:
@liggitt when will the self-labeling policy control be available as beta? We are now dependent on this, and we want to move the topology feature to beta so that overall CSI can go GA in 1.13.

liggitt commented on pkg/volume/csi/nodeinfomanager/nodeinfomanager.go:
```diff
@@ -0,0 +1,456 @@
+		}
+		return node, true, nil
+	}
+}
```
I see this adding labels to a node, what removed those labels when the CSI driver reports different labels or is removed?

What keeps track of which labels came from csi drivers and ensures stale ones are removed from the node?

verult replied:
Labels are never removed, because it could be used by multiple components, not just CSI. 

See here for more context: https://github.com/kubernetes/community/blob/7302f0a73403d493126f517cb10638f2faffc164/contributors/design-proposals/storage/container-storage-interface.md#topology-representation-in-node-objects

Here's the full design PR: https://github.com/kubernetes/community/pull/2034

verult commented:
Summary of open issues so far:
1. Kubelet is self labeling, and will be enabled by default since PluginsWatcher is moving to beta. It’s “insecure by default”. NodeRestriction updates with label prefix restrictions is in design phase.
    * For 1.12, need an *alpha* gate to control self-labeling logic. This alpha gate could be one of the following:
        * A separate `CSIDriverRegistration` flag for the entire CSI driver registration callback (including node ID annotation updates)
        * One gate for both CSI driver registration callback and CRD auto-install, which includes both `CSINodeInfo` and `CSIDriver` (the cluster registry object). This replaces the `CSICRDAutoInstall` feature.
        * One gate for all CSI features added for 1.12 (including passing workload information and not requiring attach if volume is non-attachable)
    * NodeRestriction updates is now a dependency for CSI topology. If CSI topology moves to beta, NodeRestriction has to be beta as well.
        * If we want CSI GA with topology as beta, NodeRestriction poses a constraint.
1. What should be the AddNodeInfo() behavior if CRD doesn’t exist? Continue with registration? Fail and unregister?
1. Currently we don’t control whether a user installs the CRD outside of the cluster. Is this a problem?

For (3) If CSI kubelet driver registration is enabled, and CRD auto-install is disabled, the feature will still work if an outside user installs it. The NodeRestriction plugin, if enabled (and it is by default), will limit write access to CSINodeInfo to only the node that matches the object’s name. If the plugin is explicitly disabled, anyone could write to CSINodeInfo and influence volume operations, but I think that's OK because it's an explicit disable.

@msau pointed out that even though the node ID annotation is GA, the *kubelet* logic for updating the annotation was alpha (under KubeletPluginsWatcher alpha gate). So technically we could remove the annotation logic altogether in kubelet, and if users want to rely on the annotation (for example if they are using an outdated external-provisioner), they must run the driver-registrar instead.

Given this, for (2) we can always unregister the driver since annotations will no longer exist and there’s no other driver info to fall back on.

verult commented:
Regarding which alpha gate to use, because CSICRDAutoInstall handles two CRDs, it makes sense to either:
* bundle all the CSI features under one gate, or
* have individual gates for each feature, or
* Split up CSICRDAutoInstall, then bundle everything related to CSINodeInfo into one, and everything related to CSIDriver into another

msau42 commented:
The other important decision we need to make soon is: Do we want the populating node annotations behavior to go to beta in 1.12 (current state), or pull it back to alpha and replace it entirely with CSINodeInfo

verult commented:
I'm all for replacing annotations, if possible. @saad-ali @liggitt could you confirm removing alpha logic that updates non-alpha annotation is OK?

verult commented:
If we pull back annotation logic to alpha, the remaining major decision we have to make is just: what feature gate should we use?

There is a chance that different CSI features will advance to beta and GA at different rates, so my vote is against the first option of bundling all new CSI features under one gate.

verult commented:
ACK, going to proceed with grouping features under two separate alpha flags

verult commented:
Discussed with @saad-ali offline. Since we hope to GA CSI by Q4, ideally we'd like the core logic of kubelet driver registration mechanism (including the driver cache and node ID registration) to GA by then. So our current plan is:
* Keep the current node ID annotation logic so that it could go to beta along with Plugins Watcher, so it could be GA by Q4.
* The flag for CSINodeInfo will encompass the installation logic and all of its write logic (including both node ID and topology keys), as well as Node topology labeling.
* For problem #2, the flag guarantees that the installation and write logic are bundled together. So if the CRD doesn't exist, we fail the driver registration.

verult commented:
@gnufied max attach limit rebased, also added remove logic. PTAL

verult committed:
Rearranged feature flags

verult commented:
Updated with appropriate alpha gates.

Right now each feature related to `CSIDriver` object still has its own separate feature gate. Will address that in a separate PR.

Will also address the informer issue https://github.com/kubernetes/kubernetes/issues/68378 (lacking a flag gate) in a separate PR.

@saad-ali @liggitt PTAL

verult commented:
Missing node info manager tests for when the flag gate is disabled. I'm on it.

saad-ali commented:
/lgtm
/approve

liggitt commented:
one comment on the unit test failure, authz and admission changes lgtm otherwise

it's worth noting that the removal of driver info from the node depends on observing the removal of the driver (rather than reconciling to current observed state), and doesn't handle errors in removing driver info by retrying. That doesn't necessarily block this PR, but seems important to resolve before this graduates.

verult commented:
To clarify, in this PR driver is only unregistered when there's an error adding the driver. Driver unregistration will be added in the next release. The rest of the CSI system should always assume node info is somewhat out of date, since kubelet can't provide the guarantee that updates are reflected immediately anyway, in the best case.

Does the same logic apply to driver addition as well?

verult commented on plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go:
```diff
@@ -164,6 +164,10 @@ func NodeRules() []rbacv1.PolicyRule {
 		}
 	}
+	if utilfeature.DefaultFeatureGate.Enabled(features.KubeletPluginsWatcher) {
+		csiNodeInfoRule := rbacv1helpers.NewRule("get", "create", "update", "patch", "delete").Groups("csi.storage.k8s.io").Resources("csinodeinfos").RuleOrDie()
```
That's one spot I missed. Thanks for the catch!

verult commented:
Keeping a laundry list of considerations for CSI kubelet driver registration moving forward:
https://github.com/kubernetes/kubernetes/issues/67972

liggitt commented:
/approve

verult commented:
/retest

verult commented:
The GCE e2e failure is likely caused by this PR. Taking a look

msau42 commented:
/lgtm

verult commented:
Per https://github.com/kubernetes/kubernetes/pull/68519#issuecomment-420738729: installing CRD outside k8s core brings back issue #2:
> What should be the AddNodeInfo() behavior if CRD doesn’t exist? Continue with registration? Fail and unregister?

I side with @msau42 that if the `CSINodeInfo` feature flag is enabled, one would expect topology to work, and continuing with registration by falling back to node annotation means topology breaks. So it should fail and unregister the driver. This is the current behavior so no actions required.
